samples/Config.CustomConfigurationProviders.Sample/Microsoft.Extensions.Configuration.ConfigFile/ConfigFileConfigurationProvider.cs

```diff
@@ -0,0 +1,147 @@
+    {
+        private readonly ILogger _logger;
+        private readonly string _configuration;
+        private readonly bool _loadFromFile;
+        private readonly bool _isOptional;
+
+        private readonly IEnumerable<IConfigurationParser> _parsers;
+
+        public ConfigFileConfigurationProvider(string configuration, bool loadFromFile, bool optional, params IConfigurationParser[] parsers)
+            : this(configuration, loadFromFile, optional, null, parsers)
+        { }
+
+        public ConfigFileConfigurationProvider(string configuration, bool loadFromFile, bool optional, ILogger logger, params IConfigurationParser[] parsers)
+        {
+            _loadFromFile = loadFromFile;
+            _configuration = configuration;
+            _isOptional = optional;
+            _logger = logger;
+
+            var parsersToUse = new List<IConfigurationParser> {
+                new KeyValueParser(),
+                new KeyValueParser("name", "connectionString")
+            };
+
+            parsersToUse.AddRange(parsers);
+
+            _parsers = parsersToUse.ToArray();
+        }
+
+        public override void Load()
+        {
+            if (_loadFromFile && !_isOptional && !File.Exists(_configuration))
+            {
+                throw new FileNotFoundException("Could not find configuration file to load!", _configuration);
+            }
+
+            XDocument document;
+
+            if (!TryLoadXDocument(out document))
+            {
+                return;
+            }
+
+            Data = new SortedDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            var context = new Stack<string>();
+            var dictionary = new SortedDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var child in document.Root.Elements())
+            {
+                ParseElement(child, ref context, ref dictionary);
+            }
+
+            Data = dictionary;
+        }
+
+        private bool TryLoadXDocument(out XDocument document)
+        {
+            if (!_loadFromFile)
+            {
+                document = XDocument.Parse(_configuration);
+                return true;
+            }
+
+            try
+            {
+                document = XDocument.Load(_configuration);
+                return true;
+            }
+            catch (IOException e)
+            {
+                if (_logger != null)
+                {
+                    _logger.LogError($"Unable to open {_configuration}! " + Environment.NewLine + $"Exception: {e}");
+                }
+            }
+
+            document = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Given an XElement tries to parse that element using any of the KeyValueParsers
+        /// and adds it to the results dictionary
+        /// </summary>
+        private void ParseElement(XElement element, ref Stack<string> context, ref SortedDictionary<string, string> results)
```

Comment:
I was trying to be explicit that those objects were being modified for each call. But its easy to remove.